testing/integration-tests/src/utils/tx_retries.rs

```diff
@@ -0,0 +1,51 @@
+// Copyright 2019-2024 Parity Technologies (UK) Ltd.
+// This file is dual-licensed as Apache-2.0 or GPL-3.0.
+// see LICENSE for license details.
+
+use subxt::client::OnlineClientT;
+use subxt::tx::SubmittableExtrinsic;
+use subxt::Config;
+
+pub async fn submit_tx_wait_for_finalized_success<T, C>(
+    signed_extrinsic: &SubmittableExtrinsic<T, C>,
+) -> Result<subxt::blocks::ExtrinsicEvents<T>, subxt::Error>
+where
+    T: Config,
+    C: OnlineClientT<T>,
+{
+    let submit = || async {
+        signed_extrinsic
+            .submit_and_watch()
+            .await
+            .unwrap()
+            .wait_for_finalized_success()
+            .await
+    };
+
+    const RETRY_TIME: u64 = 5;
+
+    #[cfg(lightclient)]
+    for _ in 0..2 {
+        let result = submit().await;
+
+        match result {
+            Ok(tx_in_block) => return Ok(tx_in_block),
+            Err(subxt::Error::Transaction(subxt::error::TransactionError::Dropped(_))) => {
+                tracing::info!("Transaction was dropped, retrying...");
+                // Retry if the transaction was dropped.
+                tokio::time::sleep(std::time::Duration::from_secs(RETRY_TIME)).await;
+            }
+            Err(subxt::Error::Rpc(subxt::error::RpcError::ClientError(err)))
+                if err.to_string().contains("No node available") =>
+            {
+                tracing::info!("Transaction error: {}, retrying...", err.to_string());
+
+                // Retry if the client is not connected.
+                tokio::time::sleep(std::time::Duration::from_secs(RETRY_TIME)).await;
+            }
+            Err(other) => return Err(other),
+        }
+    }
+
+    submit().await
```

Comment:
So, if using the light client, we try submitting a couple of additional times to handle light client specific issues?

I wonder if there's anything we can do to avoid needing to do this; does needing it imply that using the light client to submit transactions now is not that robust and will fail sometimes? Is there anything we can do before we try submitting that might help things (ie wait a few blocks), so that by the time we submit we don't need anything special? Is this just because of the legacy backend and everything works ok with the unstable backend (and in that case I think I'd only try testing the unstable backend with the light client)?

Reply:
Yep that makes sense. I've removed the `tx_retries` for now.
Instead maybe we should report those `tx::Dropped` events to smoldot if they still happen after an upgrade.